crates/turborepo-repository/src/change_mapper.rs

```diff
@@ -0,0 +1,233 @@
+
+        match lockfile_change {
+            Some(LockfileChange::WithContent(content)) => {
+                // if we run into issues, don't error, just assume all packages have changed
+                let Ok(lockfile_changes) = self.get_changed_packages_from_lockfile(content) else {
+                    return Ok(PackageChanges::All);
+                };
+
+                changed_pkgs.extend(lockfile_changes);
+
+                Ok(PackageChanges::Some(changed_pkgs))
+            }
+            // We don't have the actual contents, so just invalidate everything
+            Some(LockfileChange::Empty) => Ok(PackageChanges::All),
+            None => Ok(PackageChanges::Some(changed_pkgs)),
+        }
+    }
+
+    fn repo_global_file_has_changed(
+        &self,
+        default_global_deps: &[&str],
+        changed_files: &HashSet<AnchoredSystemPathBuf>,
+    ) -> Result<bool, turborepo_scm::Error> {
+        let global_deps = self.global_deps.iter().map(|s| s.as_str());
+        let filters = global_deps.chain(default_global_deps.iter().copied());
+        let matcher = wax::any(filters).unwrap();
+        Ok(changed_files.iter().any(|f| matcher.is_match(f.as_path())))
+    }
+
+    fn filter_ignored_files<'b>(
+        &self,
+        changed_files: impl Iterator<Item = &'b AnchoredSystemPathBuf> + 'b,
+    ) -> Result<HashSet<&'b AnchoredSystemPathBuf>, turborepo_scm::Error> {
+        let matcher = wax::any(self.ignore_patterns.iter().map(|s| s.as_str())).unwrap();
+        Ok(changed_files
+            .filter(move |f| !matcher.is_match(f.as_path()))
+            .collect())
+    }
+
+    // note: this could probably be optimized by using a hashmap of package paths
+    fn get_changed_packages<'b>(
+        &self,
+        files: impl Iterator<Item = &'b AnchoredSystemPathBuf>,
+        graph: &PackageGraph,
+    ) -> Result<HashSet<WorkspaceName>, turborepo_scm::Error> {
+        let mut changed_packages = HashSet::new();
+        for file in files {
+            let mut found = false;
+            for (name, entry) in graph.workspaces() {
+                if name == &WorkspaceName::Root {
+                    continue;
+                }
+                if let Some(package_path) = entry.package_json_path.parent() {
+                    if Self::is_file_in_package(file, package_path) {
+                        changed_packages.insert(name.to_owned());
+                        found = true;
+                        break;
+                    }
+                }
+            }
+            if !found {
+                // if the file is not in any package, it must be in the root package
+                changed_packages.insert(WorkspaceName::Root);
+            }
+        }
+
+        Ok(changed_packages)
+    }
+
+    fn is_file_in_package(file: &AnchoredSystemPath, package_path: &AnchoredSystemPath) -> bool {
+        file.components()
+            .zip(package_path.components())
+            .all(|(a, b)| a == b)
+    }
```

Comment:
I know this is just copied, but is there something I'm missing? Shouldn't we be able to use `==`? If they're both system paths, they should be using the same separator.

Reply:
Hmm so I think the rationale is that it accounts for cases where the paths are not of equal length. Like it just checks that the overlapping parts of those two paths are equal. But we don't check that `file` is necessarily the longer path here.